File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.ts
function maxSubarrayLength(nums: number[], k: number): number {
    const freq = new Map<number, number>();
    let i = 0, res = 0;
    for (let j = 0; j < nums.length; j++) {
        freq.set(nums[j], (freq.get(nums[j]) || 0) + 1);
        while (freq.get(nums[j])! > k) {
            freq.set(nums[i], freq.get(nums[i])! - 1);
            i++;
        }
        res = Math.max(res, j - i + 1);
    }
    return res;
}

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        int i = 0, j = 0, n = nums.size();
        int res = 0;
        while (j < n) {
            freq[nums[j]]++;
            while (freq[nums[j]] > k) {
                freq[nums[i]]--;
                i++;
            }
            res = max(res, j - i + 1);
            j++;
        }
        return res;
    }
};